onnxruntime/core/framework/debug_node_inputs_outputs_utils.cc

```diff
@@ -0,0 +1,313 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifdef DEBUG_NODE_INPUTS_OUTPUTS
+
+#include "core/framework/debug_node_inputs_outputs_utils.h"
+
+#include <iomanip>
+
+#include "core/common/path_utils.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/platform/env.h"
+
+namespace onnxruntime {
+namespace utils {
+
+namespace {
+
+bool FilterNode(const NodeDumpOptions& dump_options, const Node& node) {
+  auto match_pattern =
+      [](const std::string& value, const std::string& delimited_patterns) {
+        // match all if empty
+        if (delimited_patterns.empty()) return true;
+
+        // search for exact match against delimited patterns
+        auto pattern_begin = delimited_patterns.begin();
+        while (true) {
+          const auto pattern_end = std::find(
+              pattern_begin, delimited_patterns.end(),
+              NodeDumpOptions::FilterOptions::kPatternDelimiter);
+
+          if (std::equal(value.begin(), value.end(), pattern_begin, pattern_end)) return true;
+
+          if (pattern_end == delimited_patterns.end()) break;
+
```

Comment:
Any reason to not use unordered_set for the matching?

Reply:
I wanted to make the usage of the environment variables and direct setting of NodeDumpOptions similar. Now they both use pattern strings. A possible enhancement is to use regular expressions here.